docs/proposals/spec-reconciliation.md

```diff
@@ -0,0 +1,91 @@
+# Spec Reconciliation Proposal
+
+## Introduction
+Add a way for broker to sync the list of specs it has with registries while handling the following corner cases:
+ 1. Temporary loss of connection with registry
+ 2. A spec removed from the registry but has provisioned instance/s
+
+## Problem Description
+
+The current approach of syncing the spec from registries is as follows:
+ 1. Delete all the specs currently in the datastore
+ 2. Fetch new specs
+ 3. Add all the fetched specs to the datastore
+ 
+This leads to a bunch of bugs [Bug 1583495](https://bugzilla.redhat.com/show_bug.cgi?id=1583495), [Bug 1577810](https://bugzilla.redhat.com/show_bug.cgi?id=1583495), [Bug 1586345](https://bugzilla.redhat.com/show_bug.cgi?id=1586345)
+
+
+Instead, the broker can take the following actions:
+####Approach#1
+ 1. Fetch the specs from registry
+    1. If the fetch fails increment a failed connection counter. If the failed connection counter reaches a limit delete all the specs that do not have a service instance
+    2. If the fetch succeeds:
+        1. add new specs to the datastore
+        2. delete that specs that are removed from datastore and do not have a service instance.
+        3. mark the specs that are removed from datastore and has a service instance
+ 2. During deprovision, if the instance getting deprovisioned is the last one delete the spec from datastore
+ 3. Serve the catalog request with specs that are not marked
+  
+This will solve the problem but the task of deleting the specs is spread across time (when fetching specs and when deprovisioning)
+
+####Approach#2
+ 1. Fetch the specs from registry
+    1. If the fetch fails mark all the specs for deletion
+    2. If the fetch succeeds,
+       1. add new specs to the datastore
+       2. mark the specs that are removed from the registry.
+ 3. Run a background task to delete all the marked specs that does not have a service instance currently provisioned. The interval of running this task can be configurable
+ 4. Serve the catalog request with specs that are not marked
+ 
+This solves the problem but introduces a new parameter of configuration. 
+
+If in future, other conditions of deleting a spec crop up, just mark it for deletion and background task will take care of deleting it safely.
+
+
+Either approach will take care of the all objectives and bugs.
+
+##Other ideas
```

Comment:
My gut feeling is that these should be separate proposals and are immediately related to the reconciliation problem.